### ISSUE TITLE
[patterns] Promotion of the matched object doesn't flow from case to case, right?

### ISSUE THREAD

**Issue by eernstg** [dart-lang/language#2801]
In a matching context, the patterns in each case may be guaranteed to perform a certain amount of steps when that case fails to match (so it doesn't throw, it just proceeds to the next case). This "amount of steps" does not have to be a sequence; for instance, there could be many different ways for a `||` pattern to reach failure. But all those ways to fail might still provide a guarantee that the failing matched object can be promoted (presumably as a merge of a bunch of failure branches).

However, I would expect this analysis to be complex and brittle (change one `case` a tiny bit, and cases further down then start having various compile-time errors, or they will silently give declared variables different types, etc).

In other words, I'd actually expect that at the point where a given case _C_ is analyzed **no promotion of the matched object takes place** based on the knowledge that every earlier case than _C_ must have failed. (Note that we can still promote the matched object and various pattern-declared variables _within_ one case.)

@munificent, @stereotype441, @leafpetersen, @lrhn, @kallentu, @natebosch, @jakemac53, do you agree on this?

**Comment by lrhn** [comment]
I'd prefer to say that nothing flows from a failed match, other than caching of values.
But exhaustiveness is derived from failed matches, so *something* does come out, and we do use it for semantically significant analysis.

The question then becomes whether exhaustiveness effects cause _promotion_ in particular. I think they should, because ignoring obvious information looks stupid.
```dart
switch (someNum) {
  case double d: // d is double
  case var i; // i is int, because exhaustiveness ensures it.
}
```

Today we allow:
```dart
T? x = ...;
if (x == null) {  ...
} else {
  // x is T
}
```
I'd be sorry if switches are *less* powerful than `if` statements.

A nullable type `T?` is exhaustible by `Null` (which we can treat as an enum type exhausted by the value `null`) and `T`.
The `FutureOr<T>` type is exhaustible by `T` and `Future<T>`.
A sealed type (like `num`) is exhaustible by its immediate subtypes.

These are all (not necessarily disjoint) union types, and I think we should keep track of the remaining possible subtypes, and allow promotion of the variable when there is only one possible subtype left. 
Then the `case var i:` above works just as well as `case int i:` would, a case which is *guaranteed to match* because we've accepted that the switch is exhaustive. In an exhaustive switch, there will be at least one case which *must* match if control reaches it. If that's based on exhausting subtypes, we now know the subtype of the value.

**Comment by munificent** [comment]
> I'd actually expect that at the point where we start the analysis of a given case _C_, **no promotion of the matched object takes place** based on the knowledge that every earlier case than _C_ must have failed.

Yes, that's my understanding of the consensus we reached when we discussed this last: #2264

> The question then becomes whether exhaustiveness effects cause _promotion_ in particular. I think they should, because ignoring obvious information looks stupid.
> 
> ```dart
> switch (someNum) {
>   case double d: // d is double
>   case var i; // i is int, because exhaustiveness ensures it.
> }
> ```

My understanding is that exhaustiveness will affect promotion *after* the switch statement, but not across cases *within* it. So promotion comes into play in an example like:

```dart
num n = ...;
String? maybeString = ...;
switch (n) {
  case double d: maybeString = 'yes a string';
  case int i: maybeString = 'also a string';
}
print(maybeString.length); // OK.
```

This is example is OK because `maybeString` is promoted to a non-nullable String in every case of the switch and exhaustiveness tells us that those cases cover all values.

I do *not* think exhaustiveness should affect promotion in your example. I guess that technically makes switches dumber than if statements with regards to nullable types, but I'm OK with that. I don't want people to write code like you have there because it's extremely brittle with regards to new subtypes being added to a sealed type. I want them to write:

```dart
switch (someNum) {
  case double d: // d is double
  case int i: // i is int
}
```

This way, if a new subtype of `num` is added (unlikely, obviously, but imagine other types are used here), then they'll get an easy-to-understand "missing case" error and not a very confusing "your switch is still exhaustive but now the last case's variable has a useless type" error.

**Comment by lrhn** [comment]
We have to decide whether to treat the tests of a pattern as able to promote a variable whose value is being matched, or not. If we say "not", then things are simple, but also kind-of useless. _Less_ powerful than a current `if` statement.

Consider:
```dart
if (x case <int>[var first, ..., var last]) { 
   // is `x` promoted to `List<int>`?
}
```
It would be sad if we *couldn't* promote `x` here, where a `if (x is List<int>) { var [first, ..., last] = x; ... }` would promote.

There is a workaround, `if (x case <int>[var first, ..., var last] && var l) // l is List<int>!`, but it's not that convenient.
It is there, so we can get away with not providing promotion of the original value.
(If the intersection pattern doesn't break exhaustiveness.)

If we say that patterns never promote the original value expression, you always have to re-bind the matched value to a new variable, that's a consistent and powerful-enough choice. And it is easy to understand, easier than our current promotion.


Much easier, really. Consider:
```dart
num someNum = ...;
switch (someNum) {
  case int i: print("I'm OK");
  case double d: someNum = d.toInt();
}
// Is someNum `int` here?
```
The first case is an `int` check on the value of `someNum`. 
_Does that make `int` a type of interest for `someNum`?_ If not, we can stop here, there definitely won't be any promotion.
If `int` is a type of interest, then `someNum = d.toInt();` should promote `someNum` to `int` in that branch.
Should `case int i: ...` promote `someNum` to `int` as well? (It seems like it should, at the end of the branch, we do know that `someNum` is (still) `int`.

If we swap the cases, will it still work? (Argument for why not is that `int` is not a type of interest yet at the `someNum = d.toInt();` assignment. Very inconvenient. We might have to harvest types of interest from the entire switch before starting.)

So, yeah, maybe it's better to not go there. No promotion of a *variable* being switched on, if you need a value matching the currently matched pattern, you need to bind it. (But `<int>[var first, ...] & var list` needs to work and make `list` a `List<int>` then. Treat the matched value itself as if it is an implicit promotable variable, independently of the expression it came from. One per `case`, so no type-of-interest sharing.)

**Comment by eernstg** [comment]
@munificent wrote:
> that's my understanding of the consensus we reached when we discussed this last: https://github.com/dart-lang/language/issues/2264

Ah, I looked around, but I did not find that issue, thanks! I think your comments indicate that the following characterization is still OK:

> at the point where we start the analysis of a given case _C_, no promotion of the matched object takes place based on the knowledge that every earlier case than _C_ must have failed.

Note that this characterization is compatible with promotions that are applied to a scrutinee which is a promotable local variable after the end of the switch statement/expression. Those promotions are not affected by this issue, it's only about the static type of the scrutinee during the static analysis of each case.

@stereotype441, does this also correctly describe the actual implemented behavior? I noted that there is a diagnostic message for the following example:

```dart
void main() {
  int? maybeInt = 2 as dynamic; // No promotion.
  switch (maybeInt) {
    case == null: print('Got null');
    case var v!: print('Got an even number: ${v.isEven}'); // Info, line 5 column 18: dead_code.
  }
}
```

Checking again, this is actually _not_ a notification that the `!` check will never throw, it's a notification that the print statement will never be executed! (.. but it does actually print a line in the console at run time, so the rumors of its death are greatly exaggerated! ;-)

This means that there is some issue with the dead_code info (which is not relevant to the topic of this issue, but should still be noted), and it does look like the rule "no promotion of scrutinee for next case" is working as intended.

Finally, @lrhn, do you wish to argue in favor of any further enhancements beyond this?

**Comment by stereotype441** [comment]
In general, the philosophy I've been following when implementing flow analysis for patterns is to try to make it behave as similarly as possible to the equivalent desugared code.  In most cases, this means that no promotion information flows from case to case, because there's not much that flow analysis can conclude from a case *failing* to match.

There are a few exceptions though:

1. As of https://dart-review.googlesource.com/c/sdk/+/280861 (which hasn't landed yet), `case null` (without a guard) causes the scrutinee to be promoted to non-nullable in later cases.
2. Once I've finished adding flow analysis support for relational patterns (which I'm working on today), `case == null` will have the same behaviour.
3. Since a null-assert pattern causes an exception to be immediately thrown if the scrutinee is `null`, this is considered to immediately promote the scrutinee.  So, for example, `case (...)! when ...` causes the scrutinee to be promoted to non-nullable in later cases.
4. When a scrutinee type is a union type, and the case's required type covers one side of the union, flow analysis uses the "factor" algorithm (defined in https://github.com/dart-lang/language/blob/master/resources/type-system/flow-analysis.md) to promote the scrutinee to the remainder of the union in later cases.  So for example, if the scrutinee has a nullable type, `case Null()` promotes to non-nullable in later cases.  And if the scrutinee type is `FutureOr<R>`, then `case Future()` promotes the scrutinee to `R`, and `case R()` promotes the scrutinee to `Future<R>`.

The only union types understood by "factor" algorithm are those induced by `?` types and by `FutureOr`.  I don't plan to expand it to understand that `num == double | int`, or to support user-defined sealed types.  Exhaustiveness checking will understand about sealed types, but since it will take place after flow analysis is finished, that won't affect flow analysis.

Regarding Erik's example:
> @stereotype441, does this also correctly describe the actual implemented behavior? I noted that there is a diagnostic message for the following example:
> 
> ```dart
> void main() {
>   int? maybeInt = 2 as dynamic; // No promotion.
>   switch (maybeInt) {
>     case == null: print('Got null');
>     case var v!: print('Got an even number: ${v.isEven}'); // Info, line 5 column 18: dead_code.
>   }
> }
> ```
> 
> Checking again, this is actually _not_ a notification that the `!` check will never throw, it's a notification that the print statement will never be executed! (.. but it does actually print a line in the console at run time, so the rumors of its death are greatly exaggerated! ;-)
> 
> This means that there is some issue with the dead_code info (which is not relevant to the topic of this issue, but should still be noted), and it does look like the rule "no promotion of scrutinee for next case" is working as intended.

Yeah, this is a bug, and it's due to the fact that flow analysis doesn't yet support relational patterns.  Once I've finished adding support for relational patterns (hopefully today or tomorrow!), the dead code warning should go away, and the analyzer should warn you that the `!` is unnecessary.

**Comment by eernstg** [comment]
I think this has been clarified.
